src/python/pants/core/target_types.py

```diff
@@ -116,29 +203,52 @@ class AssetSourceField(SingleSourceField):
 
         If an http_source is provided, represents the network location to download the source from.
         The downloaded file will exist in the sandbox in the same directory as the target.
+
         `http_source` has the following signature:
+
             http_source(url: str, *, len: int, sha256: str, filename: str = "")
+
         The filename defaults to the last part of the URL path (E.g. `example.ext`), but can also be
         specified if you wish to have control over the file name. You cannot, however, specify a
         path separator to download the file into a subdirectory (you must declare a target in desired
         subdirectory).
+
         You can easily get the len and checksum with the following command:
+
             `curl -L $URL | tee >(wc -c) >(shasum -a 256) >/dev/null`
+
+        If a `per_platform` is provided, represents a mapping from platform to `http_source`, where
+        the platform is one of (linux_arm64, linux_x86_64, macos_arm64, macos_x86_64) and is
+        resolved in the execution target. Each `http_source` value MUST have the same filename provided.
         """
     )
 
     @classmethod
     def compute_value(  # type: ignore[override]
-        cls, raw_value: Optional[Union[str, HTTPSource]], address: Address
-    ) -> Optional[Union[str, HTTPSource]]:
+        cls,
+        raw_value: Optional[Union[str, http_source, per_platform[http_source]]],
+        address: Address,
+    ) -> Optional[Union[str, http_source, per_platform[http_source]]]:
         if raw_value is None or isinstance(raw_value, str):
             return super().compute_value(raw_value, address)
-        elif not isinstance(raw_value, HTTPSource):
+        elif isinstance(raw_value, per_platform):
+            raw_value.check_types(http_source)
+            value_as_dict = dataclasses.asdict(raw_value)
+            filenames = {
+                source["filename"] for source in value_as_dict.values() if source is not None
+            }
+            if len(filenames) > 1:
+                raise ValueError(
+                    "Every `http_source` in the `per_platform` must have the same `filename`,"
+                    + f"but found: {', '.join(sorted(filenames))}"
```

Comment:
whops, there'll be no space after the comma 😁

Reply:
gahh thats why softwrap exists lol

Reply:
```suggestion
                    + f" but found: {', '.join(sorted(filenames))}"
```

Reply:
> gahh thats why softwrap exists lol

Yeah, and now also encouraged over `dedent`: #17467